src/python/pants/engine/internals/graph.py

```diff
@@ -167,175 +169,291 @@ def warn_deprecated_field_type(field_type: type[Field]) -> None:
+        # The target was parametrized, and so the original Target does not exist.
+        generated = FrozenDict(
             (
+                parameterized_address,
                 target_type(
-                    generator_fields,
-                    address,
-                    name_explicitly_set=target_adaptor.name is not None,
+                    parameterized_fields,
+                    parameterized_address,
+                    name_explicitly_set=target_adaptor.name_explicitly_set,
                     union_membership=union_membership,
                 ),
-                template,
             )
-            for address, template in Parametrize.expand(address, template_fields)
-        ]
-        all_generated = await MultiGet(
-            Get(
-                GeneratedTargets,
-                GenerateTargetsRequest,
-                generate_request(
-                    generator,
-                    template_address=generator.address,
-                    template=template,
-                    overrides={
-                        name: dict(Parametrize.expand(generator.address, override))
-                        for name, override in overrides.items()
-                    },
-                ),
-            )
-            for generator, template in generators
-        )
-        parametrizations.extend(
-            _TargetParametrization(generator, generated_batch)
-            for generated_batch, (generator, _) in zip(all_generated, generators)
+            for parameterized_address, parameterized_fields in (first, *rest)
         )
+        return _TargetParametrization(None, generated)
     else:
-        first, *rest = Parametrize.expand(address, target_adaptor.kwargs)
-        if rest:
-            # The target was parametrized, and so the original Target does not exist.
-            generated = FrozenDict(
-                (
-                    parameterized_address,
-                    target_type(
-                        parameterized_fields,
-                        parameterized_address,
-                        name_explicitly_set=target_adaptor.name_explicitly_set,
-                        union_membership=union_membership,
-                    ),
+        # The target was not parametrized.
+        target = target_type(
+            target_adaptor.kwargs,
+            address,
+            name_explicitly_set=target_adaptor.name_explicitly_set,
+            union_membership=union_membership,
+        )
+        for field_type in target.field_types:
+            if (
+                field_type.deprecated_alias is not None
+                and field_type.deprecated_alias in target_adaptor.kwargs
+            ):
+                warn_deprecated_field_type(field_type)
+        return _TargetParametrization(target, FrozenDict())
+
+
+@rule
+async def _target_generator_overrides(
+    req: _TargetGeneratorOverridesRequest, unmatched_build_file_globs: UnmatchedBuildFileGlobs
+) -> _Overrides:
```

Comment:
Because this doesn't need to be independently memoized AFAICT, rather than a `@rule` it should probably be a just an async function (née `@rule_helper`), which would avoid needing to define the argument and return type wrappers.

Reply:
I seem to (to me at least) arbitrarily hit
```console
native_engine.IntrinsicError: Get(Paths, PathGlobs, PathGlobs(globs=('demo/f1.ext',), glob_match_error_behavior=<GlobMatchErrorBehavior.warn: 'warn'>, conjunction=<GlobExpansionConjunction.any_match: 'any_match'>, description_of_origin='the `overrides` field for demo:demo')) was not detected in your @rule body at rule compile time. Was the `Get` constructor called in a separate function, or perhaps dynamically? If so, it must be inlined into the @rule body.
```
when switching to a rule helper (just regular async function) here?

Reply:
Is `MultiGet` different from get in this instance?

Reply:
Hm... it shouldn't be? cc @kaos

Reply:
I can push the patch so you can see the edit.

Reply:
https://github.com/pantsbuild/pants/pull/18523/commits/0b36a0e937e929871dca8c36897af44e1de76648

Reply:
So... I _think_ that it might be a definition-order problem, because the `@rule` decorator runs during module load time? If you define the helper function before the `@rule` that calls it, does it work?

Reply:
Yes, that does work!

Reply:
Is this worth adding to the "probable issue" bit of the error?

This bit: `Was the `Get` constructor called in a separate function, or perhaps dynamically? If so, it must be inlined into the @rule body.`?

Reply:
Actually, considering your suggestions, that error seems not correct in any sense, the `Get` in this helper is dynamic afaict?

Reply:
Yea, for sure: that message hasn't been updated since rule helpers became supported.

Reply:
I opened https://github.com/pantsbuild/pants/pull/18663, not sure about the wording but imo still better than current state :man_shrugging: